src/util/expr_cast.h

```diff
@@ -140,6 +202,23 @@ auto expr_checked_cast(TExpr &base)
   return expr_dynamic_cast<T>(base);
 }
 
+/// \brief Cast a reference to a generic typet to a specific derived class.
+///   Also assert that the typet has the expected type.
+/// \tparam T The reference or const reference type to \a TUnderlying to cast to
+/// \tparam TType The original type to cast from, either typet or const typet
+/// \param base Reference to a generic \ref typet
+/// \return Reference to object of type \a T
+/// \throw std::bad_cast If \a base is not an instance of \a TUnderlying
+/// \remark If CBMC assertions (PRECONDITION) are set to abort then this will
+///   abort rather than throw if \a base is not an instance of \a TUnderlying
+template<typename T, typename TType>
+auto type_checked_cast(TType &base)
+  -> typename detail::expr_dynamic_cast_return_typet<T, TType>::type
+{
+  PRECONDITION(can_cast_type<T>(base));
```

Comment:
This looks non-optimal as `can_cast_type` is called twice (here and in `type_try_dynamic_cast`). Wouldn't it better to put the assertion on the return value of `type_try_dynamic_cast`?